### DISCUSSION TITLE
Patterning matching on *erased* functions

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#427] [Show and tell]
Everyone learns pretty early that matching on functions is verboten, because functions are *abstract*. Well, I claim this is not a problem for *erased* functions. Here is my example:

```haskell
case [] :: forall a. [a] of
  (\@a -> [] @a) -> ...
```
The lambda pattern can only bind erased variables to the left, and to the right is a normal pattern not arbitrary expression, except variables on the left can be *used* rather than *bound* on the right (which probably merits some new syntax).

I'm not saying this is worth the cost of defining this new sort of skolem (co-skolem?!?) that `a` is in the pattern. But this does unlock being able to pattern match on the empty list completely.

-----

More seriously. I doubt we need this feature, but it was useful food for thought for me so I'd thought I'd share it. I was thinking more about pattern matching like this in #400, and before that we were discussing erased variable uses in #291, where the workarounds that certainly do work there may not work here.

This also seems tangentially related to the actually-piratical use-case of being able to recur down some data type, and find that some type parameters are phantom *for the value at hand*, and getting a more general type (or more coercions) accordingly. For an example of this last bit in the real world, imagining generalizing https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5267#note_347589 for when the parameter and return types are not quite the same.

**Comment by cgibbard** [comment]
If at any point we wanted a term-level syntax to indicate the use of higher-rank types, something along these lines might do the trick.
